packages/Telephony/Actions/BatchDTMF.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+
+using AdaptiveExpressions.Properties;
+using Microsoft.Bot.Builder.Dialogs;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Components.Telephony.Actions
+{
+    /// <summary>
+    /// Aggregates input until it matches a regex pattern and then stores the result in an output property
+    /// </summary>
+    public class BatchDTMF : Dialog
+    {
+        /// <summary>
+        /// Class identifier.
+        /// </summary>
+        [JsonProperty("$kind")]
+        public const string Kind = "Microsoft.Telephony.BatchDTMF";
+        private const string AggregationDialogMemory = "dialog.aggregation";
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BatchDTMF"/> class.
+        /// </summary>
+        /// <param name="sourceFilePath">Optional, source file full path.</param>
+        /// <param name="sourceLineNumber">Optional, line number in source file.</param>
+        [JsonConstructor]
+        public BatchDTMF([CallerFilePath] string sourceFilePath = "", [CallerLineNumber] int sourceLineNumber = 0)
+            : base()
+        {
+            // enable instances of this command as debug break point
+            this.RegisterSourceLocation(sourceFilePath, sourceLineNumber);
+        }
+
+        /// <summary>
+        /// Gets or sets the regex pattern to use to decide when the dialog has aggregated the whole message.
+        /// </summary>
+        [JsonProperty("terminationConditionRegexPattern")]
+        public StringExpression TerminationConditionRegexPattern { get; set; }
+
+        /// <summary>
+        /// Gets or sets the property to assign the result to.
+        /// </summary>
+        [JsonProperty("property")]
+        public StringExpression Property { get; set; }
+
+        /// <inheritdoc/>
+        public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            //Get value of temrination string from expression
+            string regexPattern = this.TerminationConditionRegexPattern?.GetValue(dc.State);
+
+            if (dc.Context.Activity.Type != "message")
+            {
+                return new DialogTurnResult(DialogTurnStatus.Waiting);
+            }
+
+            //Is the current message the termination string?
+            if (new Regex(regexPattern).IsMatch(dc.Context.Activity.Text))
```

Comment:
This results in compiling regex every time which has a pretty intense perf implication depending on the pattern. The internal regex implementation has a cache. You cannot really keep the regex instance but what you can do is use the static overload, i.e. Regex.Match. From the[ regex docs](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex?view=net-5.0).

> The regular expression engine must compile a particular pattern before the pattern can be used. Because Regex objects are immutable, this is a one-time procedure that occurs when a Regex class constructor or a static method is called. To eliminate the need to repeatedly compile a single regular expression, the regular expression engine caches the compiled regular expressions used in static method calls. As a result, regular expression pattern-matching methods offer comparable performance for static and instance methods.

So, using the static overload you get the free caching to avoid recompilation.

Also, as a general rule you always want to pass a timeout to regex evaluation to avoid malicious inputs other RegexDenialOfService attacks. Unfortunately, I think the static overload does not have a timeout overload, if that is the case it's a 'choose your poison' scenario, but we can talk about it